### DISCUSSION TITLE
How to shut down the FastStream app after N processed RPC calls?

### DISCUSSION THREAD

**Discussion by maxalbert** [ag2ai/faststream#1221] [Q&A] [answered]
For testing purposes, I'd like to be able to shut down my FastStream app after a given number of messages have been processed.

For context: I am doing [RPC over RabbitMQ](https://faststream.airt.ai/latest/rabbit/rpc/?h=rpc), where a client app sends an RPC call and a server app returns a response. By keeping track of the number of processed messages manually via a simple counter, I can successfully shut down my client app after the Nth response has been received.

However, I'm unsure how to implement the shutdown in the _server_ app. I cannot simply shut it down from within the message handler because FastStream requires the return value in order to complete the RPC call. I was thinking of sending a signal to initiate a shutdown which would be picked up by a separate signal handler, but I don't know how I can guarantee that the response value has successfully been received by RabbitMQ so that I can safely shut down the server app. Can you advise how I might approach this? 

Many thanks!

**Comment by Lancetnik** [comment, marked as answer]
All brokers has 'gracefully_timeout' option, guarantee that all handlers will be completed before shutting down.
I think, u can setup this option and just send SIGINT signal in main thread any time (this signal triggers graceful timeout process)

**Comment by maxalbert** [comment]
Many thanks @Lancetnik! I had missed the `graceful_timeout` option. I can confirm that once I initialise the broker with a non-zero value for `graceful_timeout`, the RPC response is processed correctly and received by the client. Much appreciated!
